[EDA] Running a Settings Health Check: Date Feature Parameter

skosgi1 opened the pull request:
# Critical Changes

# Changes

# Issues Closed

# New Metadata

## Unpackaged Metadata

# Deleted Metadata

# Testing Notes

skosgi1 commented on src/classes/FeatureParameterMapper_TEST.cls:
```diff
@@ -66,18 +66,174 @@ private class FeatureParameterMapper_TEST {
+        String featureParameterName = 'SettingsHealthCheckLastRunDate';
+
+        // Set value for the Feature Parameter
+        System.FeatureManagement.setPackageDateValue(featureParameterName, featureParameterDate);
```
@MatthewBlanski wanted to check if we can set the featureParameter value here in the test method or do we want call the update updateFromFeatureParameterModel using stub and then do a getFeatureParameterModel()?

skosgi1 commented on src/classes/FeatureParameterMapper.cls:
```diff
@@ -63,7 +109,71 @@ public virtual with sharing class FeatureParameterMapper {
+    * @return TelemetryParameterName enum equivalent of the String passed
+    **************************************************************************************************************/
+    @TestVisible
+    private TelemetryParameterName convertToEnum(String toBeConvertedString) {
```
@MatthewBlanski slightly updated the conversion and error handling w.r.t to the SPEC, wanted to check with you on this if this is ok?

MatthewBlanski commented on src/lwc/healthCheck/healthCheck.js:
```diff
@@ -1,24 +1,22 @@
 
-        }
-    }
+    @wire(updateHealthCheckLastRunDate) healthCheckLastRunDate;
```
Are we using this anywhere? I don't believe we need this wire or method.

MatthewBlanski commented on unpackaged/gated/healthCheck/lwc/healthCheck/healthCheck.js:
```diff
@@ -9,16 +10,14 @@ export default class HealthCheck extends LightningElement {
 
-        }
-    }
+    @wire(updateHealthCheckLastRunDate) healthCheckLastRunDate;
```
Likewise here. I do not believe we need this.

MatthewBlanski commented on src/classes/HealthCheckVMapper.cls:
```diff
@@ -88,25 +88,31 @@ public virtual with sharing class HealthCheckVMapper {
-        FeatureParameterDateModel featureParamDateModel = 
-           (FeatureParameterDateModel)this.locateFeatureParameterService().getFeatureParameterModel();
-        String lastRunDate = featureParamDateModel.convertToString(); 
+    public virtual HealthCheckVModel getHealthCheckViewModel() {
```
I may have missed this in a previous review. This class uses a number of tabs. Please fix the spacing issues to use four spaces.

MatthewBlanski reviewed (changes requested):
There are a number of changes, some having to do with genericization and another having to do with changing a method signature that seems to be a better pattern in the code review so we can get a theoretically arbitrary feature parameter from the feature parameter service.

A few of the others have to do with LWC skilling up and how we can and cannot access methods.

skosgi1 commented on src/classes/FeatureParameterMapper_TEST.cls:
```diff
@@ -66,18 +66,150 @@ private class FeatureParameterMapper_TEST {
+            Test.stopTest();
+        } catch (FeatureParameterMapper.FeatureParameterMapperException ex) {
+            System.assert(ex.getMessage().contains('Feature Parameter'));
+            System.assert(ex.getMessage().contains('is not defined.'));
```
@MatthewBlanski would this be an ideal way to handle asserts? Do we want to rearrange the exception in main class to accommodate the entire string in 1 assert?

skosgi1 commented:
> There are a number of changes, some having to do with genericization and another having to do with changing a method signature that seems to be a better pattern in the code review so we can get a theoretically arbitrary feature parameter from the feature parameter service.
> 
> A few of the others have to do with LWC skilling up and how we can and cannot access methods.

Thanks for the review @MatthewBlanski , I have addressed the feedback, can you please give it another shot? Thx.

MatthewBlanski reviewed (changes requested):
A few more changes requested on this. We're getting closer. There are some suggestions for stubs and mocks. There are other suggestions for how to go about implementation in a manner that adds value.

MatthewBlanski commented on src/classes/FeatureParameterMapper_TEST.cls:
```diff
@@ -66,18 +66,126 @@ private class FeatureParameterMapper_TEST {
+    * @description Test method to verify that the updateFromFeatureParameterModel method updates the feature parameter date
+    *********************************************************************************************************************************************/
+     @isTest
+     private static void updateFromFeatureParameterModelPoistiveTest() {
```
A simple typo
```suggestion
     private static void updateFromFeatureParameterModelPositiveTest() {
```

MatthewBlanski commented on src/classes/HealthCheckVMapper.cls:
```diff
@@ -73,40 +73,39 @@ public virtual with sharing class HealthCheckVMapper {
-            HealthCheckDefinitionVModel healthChkDefViewModel = new HealthCheckDefinitionVModel(healthChkDefModel.name,
-                                                                                                healthChkDefModel.className,
-                                                                                                healthChkDefModel.namespace);
+            if (lastRunDate == '12/31/2015') {
+                lastRunDate = '';
+            }
```
Since Date.format() returns an arbitrary format based on the user's locale and other settings, we cannot guarantee the string output. It would suggest setting lastRunDate to an empty string. Then, if featureParamDateModel.value is not that a strongly-typed date of 12/31/2015, set lastRunDate to featureParamDateModel.convertToString().

MatthewBlanski reviewed (changes requested):
A few minor changes

MatthewBlanski commented on src/classes/HealthCheckVMapper.cls:
```diff
@@ -73,40 +73,39 @@ public virtual with sharing class HealthCheckVMapper {
+
+            return new HealthCheckVModel(lastRunDate, numberOfRuns, healthCheckDefViewModelsList);
+        }
+        catch (FeatureParameterMapper.FeatureParameterMapperException e) {
+            throw new AuraHandledException(Label.stgHealthCheckErrorLastRunDate);
+        }
```
@skosgi1 Please make a unit test for this using a stub.

MatthewBlanski commented on src/classes/HealthCheckVMapper_TEST.cls:
```diff
@@ -119,6 +119,25 @@ public class HealthCheckVMapper_TEST {
+            Test.stopTest();
+            System.assert(false, 'Should throw an exception');
+        } catch (AuraHandledException ex) {
+            System.assertEquals('System.AuraHandledException', ex.getTypeName(), 'AuraHandledException should be thrown when exception occurs.');
```
This is the right play here since the exception in this case is a passthrough with an arbitrary message.
